opm/autodiff/StandardWellsDense.hpp

```diff
@@ -1474,16 +1512,170 @@ enum WellVariablePositions {
+
+                            const Opm::PhaseUsage& pu = phase_usage_;
+
+                            if (active_[ Water ]) {
+                                aqua = well_state.wellRates()[w*np + pu.phase_pos[ Water ] ];
+                            }
+                            if (active_[ Oil ]) {
+                                liquid = well_state.wellRates()[w*np + pu.phase_pos[ Oil ] ];
+                            }
+                            if (active_[ Gas ]) {
+                                vapour = well_state.wellRates()[w*np + pu.phase_pos[ Gas ] ];
+                            }
+
+                            const int vfp        = well_controls_iget_vfp(well_control, ctrl_index);
+                            const double& thp    = well_controls_iget_target(well_control, ctrl_index);
+                            const double& alq    = well_controls_iget_alq(well_control, ctrl_index);
+
+                            // Calculating the BHP value based on THP
+                            const WellType& well_type = wells().type[w];
+                            const int first_perf = wells().well_connpos[w]; //first perforation
+
+                            if (well_type == INJECTOR) {
+                                const double dp = wellhelpers::computeHydrostaticCorrection(
+                                                  wells(), w, vfp_properties_->getInj()->getTable(vfp)->getDatumDepth(),
+                                                  wellPerforationDensities()[first_perf], gravity_);
+                                const double bhp_calculated = vfp_properties_->getInj()->bhp(vfp, aqua, liquid, vapour, thp) - dp;
+                                // apply the strictest of the bhp controlls i.e. smallest bhp for injectors
+                                if (bhp_calculated < bhp) {
+                                    bhp = bhp_calculated;
+                                }
+                            }
+                            else if (well_type == PRODUCER) {
+                                const double dp = wellhelpers::computeHydrostaticCorrection(
+                                                  wells(), w, vfp_properties_->getProd()->getTable(vfp)->getDatumDepth(),
+                                                  wellPerforationDensities()[first_perf], gravity_);
+                                const double bhp_calculated = vfp_properties_->getProd()->bhp(vfp, aqua, liquid, vapour, thp, alq) - dp;
+                                // apply the strictest of the bhp controlls i.e. largest bhp for producers
+                                if (bhp_calculated > bhp) {
+                                    bhp = bhp_calculated;
+                                }
+                            } else {
+                                OPM_THROW(std::logic_error, "Expected PRODUCER or INJECTOR type of well");
+                            }
+                        }
+                    }
+
```

Comment:
At this point, I think bhp is expected to be nonzero, as there always must be a bhp limit, correct? If so, you should add an assert() to that effect.